src/engine/effects/engineeffectsdelay.h

```diff
@@ -0,0 +1,66 @@
+#include "util/types.h"
+
+/// The effect can produce the output signal with a specific delay caused
+/// by the inner effect processing. Based on that the signal on the input
+/// of the effect does not overlap with the signal on the effect output.
+/// For using the effect in the Dry/Wet or Dry+Wet mode the dry signal
+/// and wet signal do not overlap and the effect latency is recognizable.
+///
+/// The EngineEffectsDelay offers a solution to this situation by delaying
+/// the input dry signal by a specific amount of delay. The signal delaying
+/// handles the class method EngineEffectsDelay::process. The method
+/// can be used for the one specific effect same as the effect chain if
+/// the set delay is the delay of the whole effect chain (sum of the delay
+/// of all effects in the effect chain).
+///
+/// After delaying the non-delayed signal, both signals (delayed
+/// and non-delayed) can be mixed and used together.
+class EngineEffectsDelay final : public EngineObjectConstIn {
+  public:
+    EngineEffectsDelay();
+
+    virtual ~EngineEffectsDelay(){};
+
+    /// Called from the audio thread
+
+    /// The method sets the number of frames of which will
+    /// the EngineEffectsDelay::process method delays the input signal.
+    /// By default, the EngineEffectsDelay::process method works with
+    /// a zero delay. When is the delay set, the EngineEffectsDelay::process
+    /// method works with this set delay value until the value is changed.
+    void setDelayFrames(SINT delayFrames) {
+        // Delay is reported from the effect chain by a number of frames
+        // to aware problems with a number of channels. The inner
+        // EngineEffectsDelay structure works with delay samples, so the value
+        // is recalculated for the EngineEffectsDelay usage.
+        m_delaySamples = delayFrames * mixxx::kEngineChannelCount;
+    }
+
+    /// The method delays the input buffer by the set number of samples
+    /// and returns the result in the output buffer. The input buffer
+    /// is not changed. For zero delay the input buffer is copied into
+    /// the output buffer. For non-zero delay, the output signal is returned
+    /// with the delay.
+    ///
+    /// If the number of delay samples hasn't changed, between two
+    /// EngineEffectsDelay::process method calls, the delayed signal buffers
+    /// directly follow each other as an input buffer, however, with a delay.
+    ///
+    /// If the number of delayed samples has changed between two
+    /// EngineEffectsDelay::process method calls, the new delay value is set
+    /// as actual and the output buffer is filled using cross-fading
+    /// of the presumed output buffer for the previous delay value
+    /// and of the output buffer created using the new delay value.
+    void process(const CSAMPLE* pIn, CSAMPLE* pOutput, const int iBufferSize) override;
+
+  private:
+    SINT m_delaySamples;
+    SINT m_oldDelaySamples;
+    SINT m_delayPos;
```

Comment:
Yes, I can. I used the same names as in the `EngineFilterDelay`, but I agree that names are not quite telling of their purpose.

Reply:
What do you think about this proposal:

- _m_delaySamples -> m_newDelaySamples_

  The purpose of the variable is to keep the actual setting of the number of samples saved. When the delay has changed (via `EngineEffectsDelay::setDelayFrames`), this variable is changed too and the value before the change is kept saved in the m_oldDelaySamples.

  For the `EngineEffectsDelay` usage in the `EngineEffectChain` it makes sense to use the word "new" because before every `EngineEffectsDelay::process` call the delay value is set via `EngineEffectsDelay::setDelayFrames`.

- _m_oldDelaySamples -> m_prevDelaySamples_

  The purpose of the variable is to hold the number of delay samples before the last change ("previous delay value"). This value is kept saved for use in ramping when the delay has changed. If the delay hasn't changed between two `EngineEffectsDelay::process` calls, the `m_oldDelaySamples` and `m_delaySamples` hold the same value.

- _m_delayPos -> m_delayBufferWritePos_

  The purpose of the variable is to hold the position where should be new samples written into the delay buffer (like "write pointer" for a ring buffer).

- _m_pDelayBuffer -> keep the same name_

We can discuss the clearness later while the new names will be used in the code, however, I would like to avoid multiple commits that rename variable names only.